content/blog/expressions/index.md

```diff
@@ -0,0 +1,195 @@
++++
+title = "Using Expressions Effectively in Rust"
+date = 2024-05-16
+template = "article.html"
+[extra]
+series = "Idiomatic Rust"
+resources = [
+    "[Rust, Ruby, and the Art of Implicit Returns &mdash; Earthly blog](https://earthly.dev/blog/single-expression-functions/)"
+]
++++
+
+Rust is an expression-based language, which means that almost everything in Rust is an expression.
+Functional programmers are used to the concept, but for those coming from an imperative background, it might be a bit confusing at first.
+
+Let's see how you can use expressions effectively in Rust and how they can improve your code flow.
+
+## Understanding Expressions
+
+In Rust, an expression is any construct that evaluates to a value. This includes literals, variables, function calls, blocks, and control flow statements like `if`, `match`, and `loop`. 
+
+"Everything is an expression" is a bit of an exaggeration, but it's a useful mental model while you 
+internalize the concept.
+
+Expressions return values, so they can be used in all the places where you would expect a value, like function arguments, assignments, and return statements.
+
+## A Practical Refactoring Example
+
+Here's an abstracted version of some code from a recent real-world project I was asked to refactor, keeping the main idea intact:
+
+```rust
+use std::path::{PathBuf};
+use std::fs::File;
+use std::env;
+
+/// Create a configuration file for the application at the given path
+fn create_config_file(config_path: Option<PathBuf>) -> Result<File, Box<dyn std::error::Error>> {
+    if let Some(path) = config_path {
+        return setup_config(&path);
+    } else {
+        if let Some(home_dir) = env::home_dir() {
+            let config_dir = home_dir.join(".config/my_app/");
+            return setup_config(&config_dir);
+        } else {
+            // Fallback to the current directory
+            let current_dir = PathBuf::from(".");
+            return setup_config(&current_dir);
+        }
+    }
+}
+
+fn setup_config(path: &PathBuf) -> Result<File, Box<dyn std::error::Error>> {
+    // Implementation details omitted
+    unimplemented!()
+}
+```
+
+There are a few issues with this code, but I want you to focus on the return statements
+and the duplicate logic.
+
+## Refactoring with Expressions
+
+The first observation is that we have multiple return statements in the middle of the function.
+This can make the code harder to follow and reason about.
+Let's refactor the code to have a single return statement at the end of the function.
+
+```rust
+use std::env;
+use std::fs::File;
+use std::path::PathBuf;
```

Comment:
Do we have to repeat the use statements and the `setup_config` definition in every block? I'd think once is enough.

Reply:
See my PR description above. 😉